pkg/storage/chunk/cache/embeddedcache.go

```diff
@@ -1,32 +1,338 @@
+	entriesAddedNew prometheus.Counter
+	entriesEvicted  *prometheus.CounterVec
+	entriesCurrent  prometheus.Gauge
+	totalGets       prometheus.Counter
+	totalMisses     prometheus.Counter
+	staleGets       prometheus.Counter
+	memoryBytes     prometheus.Gauge
+}
+
+type cacheEntry struct {
+	updated time.Time
+	key     string
+	value   []byte
+}
+
 // EmbeddedCacheConfig represents in-process embedded cache config.
 type EmbeddedCacheConfig struct {
-	Enabled   bool          `yaml:"enabled,omitempty"`
-	MaxSizeMB int64         `yaml:"max_size_mb"`
-	TTL       time.Duration `yaml:"ttl"`
+	Enabled      bool          `yaml:"enabled,omitempty"`
+	MaxSizeMB    int64         `yaml:"max_size_mb"`
+	MaxSizeItems int           `yaml:"max_size_items"`
+	TTL          time.Duration `yaml:"ttl"`
 
 	// PurgeInterval tell how often should we remove keys that are expired.
-	// by default it takes `DefaultPurgeInterval`
+	// by default it takes `defaultPurgeInterval`
 	PurgeInterval time.Duration `yaml:"-"`
 }
 
 func (cfg *EmbeddedCacheConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
 	f.BoolVar(&cfg.Enabled, prefix+"embedded-cache.enabled", false, description+"Whether embedded cache is enabled.")
 	f.Int64Var(&cfg.MaxSizeMB, prefix+"embedded-cache.max-size-mb", 100, description+"Maximum memory size of the cache in MB.")
+	f.IntVar(&cfg.MaxSizeItems, prefix+"embedded-cache.max-size-items", 0, description+"Maximum number of entries in the cache.")
 	f.DurationVar(&cfg.TTL, prefix+"embedded-cache.ttl", time.Hour, description+"The time to live for items in the cache before they get purged.")
-
 }
 
 func (cfg *EmbeddedCacheConfig) IsEnabled() bool {
 	return cfg.Enabled
 }
+
+// NewEmbeddedCache returns a new initialised EmbeddedCache.
+func NewEmbeddedCache(name string, cfg EmbeddedCacheConfig, reg prometheus.Registerer, logger log.Logger, cacheType stats.CacheType) *EmbeddedCache {
+	if cfg.MaxSizeMB == 0 && cfg.MaxSizeItems == 0 {
+		// zero cache capacity - no need to create cache
+		level.Warn(logger).Log("msg", "neither embedded-cache.max-size-mb nor embedded-cache.max-size-items is set", "cache", name)
+		return nil
+	}
+
+	// Set a default interval for the ticker
+	// This can be overwritten to a smaller value in tests
+	if cfg.PurgeInterval == 0 {
+		cfg.PurgeInterval = defaultPurgeInterval
+	}
+
+	cache := &EmbeddedCache{
+		cacheType: cacheType,
+
+		maxSizeItems: cfg.MaxSizeItems,
+		maxSizeBytes: uint64(cfg.MaxSizeMB * 1e6),
+		entries:      make(map[string]*list.Element),
+		lru:          list.New(),
+
+		done: make(chan struct{}),
+
+		entriesAdded: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace:   "querier",
+			Subsystem:   "cache",
+			Name:        "added_total",
+			Help:        "The total number of Put calls on the cache",
+			ConstLabels: prometheus.Labels{"cache": name},
+		}),
+
+		entriesAddedNew: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace:   "querier",
```

Comment:
I wonder if we should take this opportunity to correct these metrics, too.
All our metrics start with `loki_` (well... except the ones that start with `cortex_`), but these don't. It also feels weird to limit the scope of this to the querier since many components can use this cache. Again, not necessarily for this PR 👍

Reply:
agree, was talking to @chaudum about renaming these. i'll raise a separate pr